Firestore/Example/Tests/Integration/FSTTransactionTests.mm

```diff
@@ -20,184 +20,346 @@
+    [expectation fulfill];
+  }];
+  [_testCase awaitExpectations];
+
+  [self cleanupTester];
+}
+
+- (void)expectNoDoc {
+  [self prepareDoc];
+  [self runSuccessfulTransaction];
+
+  XCTestExpectation *expectation = [_testCase expectationWithDescription:@"get"];
+  [_docRef getDocumentWithCompletion:^(FIRDocumentSnapshot *snapshot, NSError *error) {
+    [self->_testCase assertDoesNotExistWithSnapshot:snapshot error:error];
+    [expectation fulfill];
+  }];
+  [_testCase awaitExpectations];
+
+  [self cleanupTester];
+}
+
+- (void)expectError:(FIRFirestoreErrorCode)expected {
+  [self prepareDoc];
+  [self runFailingTransaction:expected];
+
+  [self cleanupTester];
+}
+
+- (void)prepareDoc {
+  _docRef = [[self->_db collectionWithPath:@"nonexistent"] documentWithAutoID];
+  if (_fromExistingDoc) {
+    [_docRef setData:@{@"foo" : @"bar"}];
+
+    XCTestExpectation *expectation = [_testCase expectationWithDescription:@"get"];
+
+    [_docRef getDocumentWithCompletion:^(FIRDocumentSnapshot *snapshot, NSError *error) {
+      [self->_testCase assertExistsWithSnapshot:snapshot error:error];
+      [expectation fulfill];
+    }];
+
+    [_testCase awaitExpectations];
+  }
+}
+
+- (void)runSuccessfulTransaction {
+  XCTestExpectation *expectation = [_testCase expectationWithDescription:@"runTransaction"];
+  [_db
       runTransactionWithBlock:^id _Nullable(FIRTransaction *transaction, NSError **error) {
-        [transaction getDocument:doc error:error];
-        XCTAssertNil(*error);
+        for (TransactionStage stage in self->_stages) {
+          stage(transaction, self->_docRef);
+        }
         return @YES;
       }
       completion:^(id _Nullable result, NSError *_Nullable error) {
-        XCTAssertNil(result);
-        // We currently require every document read to also be written.
-        // TODO(b/34879758): Fix this check once we drop that requirement.
-        XCTAssertNotNil(error);
-        XCTAssertEqual(error.code, FIRFirestoreErrorCodeInvalidArgument);
         [expectation fulfill];
+        NSString *message =
+            [NSString stringWithFormat:@"Expected the sequence %@, to succeed, but got %ld.",
+                                       [self listStages], [error code]];
+        [self->_testCase assertNilErrorWithMessage:message error:error];
       }];
-  [self awaitExpectations];
-}
 
-- (void)testDeleteDocument {
-  FIRFirestore *firestore = [self firestore];
-  FIRDocumentReference *doc = [[firestore collectionWithPath:@"towns"] documentWithAutoID];
-  [self writeDocumentRef:doc data:@{@"foo" : @"bar"}];
-  FIRDocumentSnapshot *snapshot = [self readDocumentForRef:doc];
-  XCTAssertEqualObjects(@"bar", snapshot[@"foo"]);
+  [_testCase awaitExpectations];
+}
 
-  XCTestExpectation *expectation = [self expectationWithDescription:@"transaction"];
-  [firestore
+- (void)runFailingTransaction:(FIRFirestoreErrorCode)expected {
```

Comment:
Objective-C method names need to incorporate their parameters. See [Apple's Cocoa Guidelines](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/CodingGuidelines/Articles/NamingMethods.html) for more details.

In this case it would be something like `runFailingTransactionWithError:`.